**Numerical Linear Algebra for Financial Engineering**

**Homework 1**

**Jose Ferreira**

----------------
**(1) Find the value of $x$ such that the matrix below has rank 3:**
$$
   \begin{pmatrix} 
      x & 1 & 1 & 1 \\ 
      1 & x & 1 & 1 \\ 
      1 & 1 & x & 1 \\
      1 & 1 & 1 & x 
   \end{pmatrix}
$$

Let M be the matrix above. Calculate the determinant of M by applying row reducing operations until the matrix is upper triangular. 
For the rank to be 3 (or less), the determinant needs to be equal to zero.
$$
\begin{align}
det(M)=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      x & 1 & 1 & 1 \\ 
      1 & 1 & x & 1 \\
      1 & 1 & 1 & x 
   \end{vmatrix} &[swap(r1,r2)]\\ 
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      1 & 1 & x & 1 \\
      1 & 1 & 1 & x 
   \end{vmatrix} &[r2 = r2 - x*r1]\\
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 1-x & x-1 & 0 \\
      1 & 1 & 1 & x 
   \end{vmatrix} &[r3 = r3 - r1]\\
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 1-x & x-1 & 0 \\
      0 & 1-x & 0 & x-1 
   \end{vmatrix} &[r4 = r4 - r1]\\
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 0 & \frac{(x-1)(2+x)}{1+x} & \frac{x-1}{1+x} \\
      0 & 1-x & 0 & x-1 
   \end{vmatrix} &[r3 = r3 -\frac{1-x}{1-x^2}*r2]\\   
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 0 & \frac{(x-1)(2+x)}{1+x} & \frac{x-1}{1+x} \\
      0 & 0 & \frac{x-1}{1+x} & \frac{(x-1)(2+x)}{1+x} 
   \end{vmatrix} &[r4 = r4 -\frac{1-x}{1-x^2}*r2]\\   
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 0 & \frac{(x-1)(2+x)}{1+x} & \frac{x-1}{1+x} \\
      0 & 0 & 0 & \frac{(x-1)(3+x)}{2+x} 
   \end{vmatrix} &[r4 = r4 -\frac{1}{2+x}*r3]\\      
=& -\frac{(1-x^2)(x-1)(2+x)(x-1)(3+x)}{(1+x)(2+x)}\\
=& (x-1)^3(3+x)
\end{align}
$$

Therefore, in order for the matrix to have a rank of 3 or less, $x$ must be $1$  or $-3$. If $x = 1$, the matrix is composed of only ones and its rank would be $1$. If $x = -3$:

$$
\begin{align}
M=&   \begin{pmatrix} 
      -3 & 1 & 1 & 1 \\ 
      1 & -3 & 1 & 1 \\ 
      1 & 1 & -3 & 1 \\
      1 & 1 & 1 & -3 
   \end{pmatrix}\\
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      -3 & 1 & 1 & 1 \\ 
      1 & 1 & -3 & 1 \\
      1 & 1 & 1 & -3 
   \end{pmatrix} &[swap(r1,r2)]\\ 
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      1 & 1 & -3 & 1 \\
      1 & 1 & 1 & -3 
   \end{pmatrix} &[r2=r2+3r1]\\ 
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 4 & -4 & 0 \\
      1 & 1 & 1 & -3 
   \end{pmatrix} &[r3=r3-r1]\\ 
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 4 & -4 & 0 \\
      0 & 4 & 0 & -4 
   \end{pmatrix} &[r4=r4-r1]\\ 
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 0 & -2 & 2 \\
      0 & 4 & 0 & -4 
   \end{pmatrix} &[r3=r3+\frac{1}{2}r2]\\
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 0 & -2 & 2 \\
      0 & 0 & 2 & -2 
   \end{pmatrix} &[r4=r4+\frac{1}{2}r2]\\   
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 0 & -2 & 2 \\
      0 & 0 & 0 & 0 
   \end{pmatrix} &[r4=r4+r3]\\   
\iff& rank(M) = 3 
\end{align}
$$

-------------
**(2) Show that the product of two symmetric matrices is not necessarily a symmetric matrix.
Hint: Find 2x2 matrices $A$ and $B$ such that $A=A^t$, $B=B^t$, and $(AB)^t  \neq AB$**

By counterexample, let A and B be matrices of the form:
    $$
    A=\begin{pmatrix} 
      a & a \\ 
      a & -a
   \end{pmatrix}\\
   B=\begin{pmatrix} 
      0 & a \\ 
      a & -a
   \end{pmatrix}\\
    $$

Both A and B are symmetric matrices but:
$$
AB=\begin{pmatrix} 
      a^2 & 0 \\ 
      -a^2 & 2a^2
   \end{pmatrix}\\
$$
is not symmetric

---------------
**(3) Let $A$ be a square matrix of size n. Show that $AD=DA$ for any diagonal matrix D of size n if and only if $A$ is a diagonal matrix.**

Let $D = diag(d_{k})_{k=1:n}$ and $A = col(a_{k})_{k=1:n}$, then $AD = col(a_{k}d_{k}) \iff (AD)(j,k) = d_{k}A(j,k)$

Also, if $A = row(r_{j})_{j=1:n}$, then $DA = row(d_{j}r_{j})_{j=1:n} \iff (DA)(j,k) = d_{j}A(j,k)$

Assuming all the diagonal entries in D are distinct: 

$$
\begin{align}
AD = DA 
\iff& d_{k}A(j,k) = d_{j}A(j,k)\\ 
\iff& (d_{k} - d_{j})A(j,k) = 0, \forall 1\leq j \neq k \leq n\\
\iff& A(j,k) = 0, \forall 1\leq j \neq k \leq n\\
\iff& A = diag(a_{k})_{k=1:n}
\end{align}
$$

If the diagonal entries of D are not all distinct, some non-diagonal entries of A could be non-zero and the commutative property will also be preserved.

-------------
**(4) Let $M_1$, $M_2$, $M_3$ and $M_4$ be square matrices of the same size such that $M_1M_2M_3M_4 = I$.
Show that
$$M_1M_2M_3M_4 = M_2M_3M_4M_1 = M_3M_4M_1M_2 = M_4M_1M_2M_3 = I$$**

$$
\begin{align}
M_1M_2M_3M_4 = I \implies& M_1(M_2M_3M_4) = I\\
\implies& (M_2M_3M_4) = M_1^{-1}\\
\implies& (M_2M_3M_4)M_1 = M_2M_3M_4M_1 = M_1^{-1}M_1 = I\\
\end{align}
$$

$$
\begin{align}
M_1M_2M_3M_4 = I \implies& (M_1M_2)(M_3M_4) = I\\
\implies& (M_3M_4) = (M_1M_2)^{-1}\\
\implies& (M_3M_4)(M_1M_2) = M_3M_4M_1M_2 = (M_1M_2)^{-1}(M_1M_2) = I\\
\end{align}
$$

$$
\begin{align}
M_1M_2M_3M_4 = I \implies& (M_1M_2M_3)M_4 = I\\
\implies& M_4 = (M_1M_2M_3)^{-1}\\
\implies& (M_4)(M_1M_2M_3) = M_4M_1M_2M_3 = (M_1M_2M_3)^{-1}(M_1M_2M_3) = I\\
\end{align}
$$

---------
**(5) If A is a nonsingular matrix, show that
$$(I+A)^{-1} = I - (I+A^{-1})^{-1}$$**

$$
\begin{align}
(I+A)^{-1} = I - (I+A^{-1})^{-1}\\ 
\implies& (I - (I+A^{-1})^{-1})(I+A) = I &\text{[def. inverse matrix]}\\
\implies& I - (I+A^{-1})^{-1} + A - (I+A^{-1})^{-1}A = I &\text{[distributive property]}\\
\implies& I - A[(I+A^{-1})^{-1}A^{-1} + I - (I+A^{-1})^{-1}] = I &\text{[factor A]}\\
\implies& I - A[(I+A^{-1})^{-1}(I+A^{-1}) - I] = I &\text{[factor } (I+A)^{-1} \text{]} \\
\implies& I - A[I - I] = I &\text{[simplify]} \\
\implies& I = I &\text{[simplify]} \\
\implies& TRUE \\
\end{align}
$$

-------------
**(6) Show that the $n \times n$ matrix $A$ has rank 1 if and only if there exists an $n \times 1$ nonzero vector $v$ such that $A=vv^t$** 

If $A=vv^t$, all rows are a linear combination of vector $v$ and, therefore, $rank(A) = 1$. Also, note that a vector has by definition a rank equal to 1, multiplying two vectors must result in a matrix of rank equal to 1. 

$$
\begin{align}
A=vv^t =& (vv^t)(j,k) = v_{j}v^t_{k}, \forall j =1:n, k=1:n\\
\implies& Ax = (v_{j}row(r_{j}))x, \forall j =1:n\\
\implies& Ax = v_{j}(row(r_{j})x), \forall j =1:n\\
\implies& \text{All rows are linearly dependent}\\
\implies& rank(A) = 1
\end{align}
$$

If $rank(A) = 1$, there exists exactly one eigenvector $v$ such that $Av = \lambda v$ and, assuming $v$ is a nonzero vector, the matrix can be expressed as $A = vv^t$ since any vector $w$ such that $A = vw^t$ (possible since rank(A) = 1) can be written as $w = \lambda v$ if  $\vert\vert{v}\vert\vert=\sqrt{\lambda}$.      

$$
\begin{align}
rank(A) = 1\\
\implies& Au = kv, \forall u \in \mathbb R^n \\
\implies& A = v(col(w_{k})), \forall k =1:n\\
\implies& A = vw^t\\    
\implies& A = v_*(v_*^t), v_*=\sqrt{\lambda}v \text{ since } Av = \lambda v\\
\end{align}
$$


-------------
**(7) Let $x$ and $y$ be column vectors of size $n$, and let $I$ be the identity matrix of size $n$. Show that the matrix $I - xy^t$ is nonsingular if and only if $y^tx \neq 1$.**

Let $A = (I-xy^t)$ and assume $y^tx \neq 1$:

$$
\begin{align}
(I-xy^t)A^{-1} = I\\
\implies& A^{-1} - xy^tA^{-1} = I\\
\implies& y^tA^{-1} - y^txy^tA^{-1} = y^t &\text{[multiply by $y^t$]}\\
\implies& (1 - y^tx)(y^tA^-1) = y^t\\
\implies& (1 - y^tx)(y^tA^-1) = y^t\\
\implies& A^-1 = I + \frac{xy^t}{(1 - y^tx)}\\
\end{align}
$$
Since $y^tx \neq 1$, the inverse of matrix A is well defined.  
If $y^tx = 1$, the inverse does not exist.

Furthermore, assuming $y^tx = 1$:
$$
\begin{align}
(I-xy^t)x =& x - xy^tx\\
=& x - x(y^tx)\\
=&0\\
\implies& \text{The matrix is singular if $y^tx = 1$}
\end{align}
$$

Therefore, $I - xy^t$ is nonsingular $\iff y^tx \neq 1$  

---------------
**(8) The covariance matrix of five random variables is
$$
\Sigma =
\begin{pmatrix} 
      1 & -0.525 & 1.375 & -0.075 & -0.75 \\ 
      -0.525 & 4 & 0.1875 & 0.1875 & -0.675 \\ 
      1.375 & 0.1875 & 12.25 & 0.4375 & -1.875 \\ 
      -0.075 & 0.1875 & 0.4375 & 6.25 & 0.3 \\ 
      -0.75 & -0.675 & -1.875 & 0.3 & 4.41
   \end{pmatrix}\\
$$
Find the correlation matrix of these random variables.** 

The correlation matrix $\Omega$ is given by
$$\Omega = (D_{\sigma})^{-1}\Sigma(D_{\sigma})^{-1}$$ 
where $D_{\sigma}$ is the diagonal matrix with diagonal values equivalent to the standard deviations of each random variable $X_{i}$  
In this case, 
$$
\begin{align}
D_{\sigma} =& 
\begin{pmatrix} 
      \sqrt{1} & 0 & 0 & 0 & 0 \\ 
      0 & \sqrt{4} & 0 & 0 & 0 \\ 
      0 & 0 & \sqrt{12.25} & 0 & 0 \\ 
      0 & 0 & 0 & \sqrt{6.25} & 0 \\ 
      0 & 0 & 0 & 0 & \sqrt{4.41}
   \end{pmatrix}\\
=& 
\begin{pmatrix} 
      1 & 0 & 0 & 0 & 0 \\ 
      0 & 2 & 0 & 0 & 0 \\ 
      0 & 0 & 3.5 & 0 & 0 \\ 
      0 & 0 & 0 & 2.5 & 0 \\ 
      0 & 0 & 0 & 0 & 2.1
\end{pmatrix}\\
\end{align}
$$

$$
\begin{align}
(D_{\sigma})^{-1} 
=& diag\left(\frac{1}{d_{k}}\right)_{k=1:n}\\
=& 
\begin{pmatrix} 
      1 & 0 & 0 & 0 & 0 \\ 
      0 & \frac{1}{2} & 0 & 0 & 0 \\ 
      0 & 0 & \frac{2}{7} & 0 & 0 \\ 
      0 & 0 & 0 & \frac{2}{5} & 0 \\ 
      0 & 0 & 0 & 0 & \frac{10}{21}
\end{pmatrix}\\
\end{align}
$$

In [88]:
sigma = [1 -0.525 1.375 -0.075 -0.75; 
         -0.525 4 0.1875 0.1875 -0.675;
         1.375 0.1875 12.25 0.4375 -1.875;
         -0.075 0.1875 0.4375 6.25 0.3;
         -0.75 -0.675 -1.875 0.3 4.41]
d_sigma_inverse = inv(Diagonal([sqrt(sigma[1,1]); 
                                sqrt(sigma[2,2]);
                                sqrt(sigma[3,3]);
                                sqrt(sigma[4,4]);
                                sqrt(sigma[5,5])]))
omega = d_sigma_inverse*sigma*d_sigma_inverse

5×5 Array{Float64,2}:
  1.0       -0.2625      0.392857   -0.03       -0.357143 
 -0.2625     1.0         0.0267857   0.0375     -0.160714 
  0.392857   0.0267857   1.0         0.05       -0.255102 
 -0.03       0.0375      0.05        1.0         0.0571429
 -0.357143  -0.160714   -0.255102    0.0571429   1.0      

$$
\begin{align}
\Omega
=& 
\begin{pmatrix} 
1.0 & -0.2625 & 0.392857 & -0.03 & -0.357143 \\ 
-0.2625 & 1.0 & 0.0267857 & 0.0375 & -0.160714 \\
0.392857 & 0.0267857 & 1.0 & 0.05 & -0.255102 \\
-0.03 & 0.0375 & 0.05 & 1.0 & 0.0571429 \\
-0.357143 & -0.160714 & -0.255102 & 0.0571429 & 1.0 \end{pmatrix}\\
\end{align}
$$


----------
**(9) The correlation matrix of five random variables is
$$
\begin{align}
\Omega
=& 
\begin{pmatrix} 
1 & -0.25 & 0.15 & -0.05 & -0.3 \\ 
-0.25 & 1 & -0.1 & -0.25 & 0.1\\
0.15 & -0.1 & 1 & 0.2 & 0.05 \\
-0.05 & -0.25 & 0.2 & 1 & 0.1 \\
-0.3 & 0.1 & 0.05 & 0.1 & 1   
\end{pmatrix}\\
\end{align}
$$
(i) Compute the covariance matrix of these random variables if their standard deviations are 0.1, 0.2, 0.5, 1 and 2, in this order.**

The covariance matrix, $\Sigma$ is given by the expression $$\Sigma = D_{\sigma}\Omega D_{\sigma}$$ where 
$$
\begin{align}
D_{\sigma} =& diag\left(\sigma_{k}\right)_{k=1:n}\\
=&
\begin{pmatrix} 
      0.1 & 0 & 0 & 0 & 0 \\ 
      0 & 0.2 & 0 & 0 & 0 \\ 
      0 & 0 & 0.5 & 0 & 0 \\ 
      0 & 0 & 0 & 1 & 0 \\ 
      0 & 0 & 0 & 0 & 2
\end{pmatrix}\\
\end{align}
$$ 


In [8]:
omega = [1 -0.25 0.15 -0.05 -0.3; 
         -0.25 1 -0.1 -0.25 0.1;
         0.15 -0.1 1 0.2 0.05;
         -0.05 -0.25 0.2 1 0.1;
         -0.3 0.1 0.05 0.1 1]
d_sigma = Diagonal([0.1; 0.2; 0.5; 1; 2])
sigma = d_sigma*omega*d_sigma

5×5 Array{Float64,2}:
  0.01    -0.005   0.0075  -0.005  -0.06
 -0.005    0.04   -0.01    -0.05    0.04
  0.0075  -0.01    0.25     0.1     0.05
 -0.005   -0.05    0.1      1.0     0.2 
 -0.06     0.04    0.05     0.2     4.0 

$$
\Sigma =
\begin{pmatrix} 
      0.01 & -0.005 & 0.0075 & -0.005 & -0.06 \\ 
      -0.005 & 0.04 & -0.01 & -0.05 & 0.04 \\ 
      0.0075 & -0.01 & 0.25 & 0.1 & 0.05 \\ 
      -0.005 & -0.05 & 0.1 & 1 & 0.2 \\ 
      -0.06 & 0.04 & 0.05 & 0.2 & 4
   \end{pmatrix}\\
$$

**(ii) Compute the covariance matrix of these random variables if their standard deviations are 2, 1, 0.5, 0.2, and 0.1, in this order.**

In [9]:
omega = [1 -0.25 0.15 -0.05 -0.3; 
         -0.25 1 -0.1 -0.25 0.1;
         0.15 -0.1 1 0.2 0.05;
         -0.05 -0.25 0.2 1 0.1;
         -0.3 0.1 0.05 0.1 1]
d_sigma = Diagonal([2; 1; 0.5; 0.2; 0.1])
sigma = d_sigma*omega*d_sigma

5×5 Array{Float64,2}:
  4.0   -0.5    0.15    -0.02   -0.06  
 -0.5    1.0   -0.05    -0.05    0.01  
  0.15  -0.05   0.25     0.02    0.0025
 -0.02  -0.05   0.02     0.04    0.002 
 -0.06   0.01   0.0025   0.002   0.01  

$$
\Sigma =
\begin{pmatrix} 
      4 & -0.5 & 0.15 & -0.02 & -0.06 \\ 
      -0.5 & 1 & -0.05 & -0.05 & 0.01 \\ 
      0.15 & -0.05 & 0.25 & 0.2 & 0.0025 \\ 
      -0.02 & -0.05 & 0.02 & 0.04 & 0.002 \\ 
      -0.06 & 0.01 & 0.0025 & 0.002 & 0.01
   \end{pmatrix}\\
$$

------------
**(10) The file *indeces-close-jan3-jan31-2017.xslx* contains the January 3, 2017 - January 31, 2017 end of day values of Dow Jones, Nasdaq, and S&P 500.**

**(i) Compute the log daily returns of the three indices over the given time period.** 

The log returns for each index are calculated as the difference between the natural logarithm of the (i+1)-th period and the i-th period.

In [49]:
using ExcelReaders
using DataFrames

df = readxlsheet(DataFrame, "indeces-close-jan3-jan31-2017.xlsx", "201701_Indices");

In [22]:
dow_log_returns = diff(log(df[:DOW]))

19-element DataArrays.DataArray{Float64,1}:
  0.00303337 
 -0.00215209 
  0.00323667 
 -0.00383527 
 -0.00160288 
  0.0049611  
 -0.00317625 
 -0.000264955
 -0.00296939 
 -0.00111269 
 -0.00365835 
  0.00479528 
 -0.00138291 
  0.00568393 
  0.00779364 
  0.00161319 
 -0.000354816
 -0.00612251 
 -0.0053742  

In [13]:
nasdaq_log_returns = diff(log(df[:NASDAQ]))

19-element DataArrays.DataArray{Float64,1}:
  0.0087878  
  0.00199544 
  0.00601693 
  0.00194696 
  0.00360893 
  0.00212858 
 -0.00290873 
  0.00478886 
 -0.00636925 
  0.00305018 
 -0.00280646 
  0.00274889 
 -0.000430335
  0.0086105  
  0.00983901 
 -0.000205042
  0.000989684
 -0.00834984 
  0.000192382

In [23]:
sp_log_returns = diff(log(df[:SP500]))

19-element DataArrays.DataArray{Float64,1}:
  0.00570596 
 -0.000770968
  0.00351079 
 -0.00355491 
  0.0        
  0.00282564 
 -0.00214711 
  0.00184813 
 -0.00297191 
  0.0017622  
 -0.00361584 
  0.00336058 
 -0.00269375 
  0.00654314 
  0.00799405 
 -0.000735655
 -0.00086684 
 -0.00602767 
 -0.000890302

**(ii) Compute the sample covariance matrix of the log daily returns of the three indices over the given time period.**

We construct the normalized matrix $T_{x}$ by substracting the mean from the log returns vectors and calculate the covariance matrix using formula (1.35) in the textbook: $$\Sigma_{x} = \frac{1}{N-1}(T_{x})^tT_{x}$$

In [31]:
dow_norm_log_returns = dow_log_returns - mean(dow_log_returns)
nasdaq_norm_log_returns = nasdaq_log_returns - mean(nasdaq_log_returns)
sp_norm_log_returns = sp_log_returns - mean(sp_log_returns)
T_norm = hcat(dow_norm_log_returns,nasdaq_norm_log_returns,sp_norm_log_returns)
sigma = size(T_norm,1)*transpose(T_norm)*T_norm

3×3 DataArrays.DataArray{Float64,2}:
 0.00544932  0.00500235  0.00465467
 0.00500235  0.00793864  0.00563904
 0.00465467  0.00563904  0.0049076 

**(iii) Compute the percentage daily returns of the three indices over the given time period.**

The percentage daily returns are calculated as $\frac{S_{i+1}-S_{i}}{S_{i}}, \forall i=2:n$ 

In [45]:
dow_returns = diff(df[:DOW])./df[2:end,:DOW]

19-element DataArrays.DataArray{Float64,1}:
  0.00302878 
 -0.0021544  
  0.00323143 
 -0.00384263 
 -0.00160417 
  0.00494881 
 -0.0031813  
 -0.000264991
 -0.0029738  
 -0.00111331 
 -0.00366505 
  0.0047838  
 -0.00138387 
  0.0056678  
  0.00776335 
  0.00161189 
 -0.000354879
 -0.00614129 
 -0.00538867 

In [46]:
nasdaq_returns = diff(df[:NASDAQ])./df[2:end,:NASDAQ]

19-element DataArrays.DataArray{Float64,1}:
  0.0087493  
  0.00199345 
  0.00599887 
  0.00194507 
  0.00360242 
  0.00212632 
 -0.00291297 
  0.00477741 
 -0.00638958 
  0.00304553 
 -0.0028104  
  0.00274511 
 -0.000430427
  0.00857353 
  0.00979076 
 -0.000205063
  0.000989194
 -0.0083848  
  0.000192363

In [47]:
sp_returns = diff(df[:SP500])./df[2:end,:SP500]

19-element DataArrays.DataArray{Float64,1}:
  0.00568972 
 -0.000771265
  0.00350463 
 -0.00356123 
  0.0        
  0.00282165 
 -0.00214942 
  0.00184643 
 -0.00297633 
  0.00176065 
 -0.00362238 
  0.00335494 
 -0.00269738 
  0.00652178 
  0.00796219 
 -0.000735925
 -0.000867216
 -0.00604588 
 -0.000890698

**(iv) Compute the sample covariance matrix of the percentage daily returns of the three indices over the given time period.**

In the same way as in point (ii), we construct the normalized matrix $T_{x}$ and calculate the covariance matrix using formula (1.35) in the textbook.

In [48]:
dow_norm_returns = dow_returns - mean(dow_returns)
nasdaq_norm_returns = nasdaq_returns - mean(nasdaq_returns)
sp_norm_returns = sp_returns - mean(sp_returns)
T_norm_2 = hcat(dow_norm_returns,nasdaq_norm_returns,sp_norm_returns)
sigma = size(T_norm_2,1)*transpose(T_norm_2)*T_norm_2

3×3 DataArrays.DataArray{Float64,2}:
 0.00544175  0.00499114  0.00464533
 0.00499114  0.00792014  0.00562336
 0.00464533  0.00562336  0.00489637

--------------
**(11) The file *indices-july2016.csv* contains the January 2016 - July 2016 end of day values of nine major US indices.**

**(i) Compute the sample covariance matrix of the daily percentage returns of the indices, and the corresponding sample correlation matrix. Compute the sample covariance and correlation matrices for daily log returns, and compare them with the corresponding matrices for daily percentage returns.**

The covariance matrix is calculated using the process described in problem 10. 

In [98]:
index_data = readtable("indices-july2016.csv");
dfmt = Dates.DateFormat("m/d/y")
index_data[:Date] = Date(index_data[:Date], dfmt)

daily_returns = hcat(diff(index_data[:BAC])./index_data[2:end,:BAC], 
                 diff(index_data[:BNS])./index_data[2:end,:BNS],
                 diff(index_data[:C])./index_data[2:end,:C],
                 diff(index_data[:GS])./index_data[2:end,:GS],
                 diff(index_data[:JPM])./index_data[2:end,:JPM],
                 diff(index_data[:MS])./index_data[2:end,:MS],
                 diff(index_data[:TD])./index_data[2:end,:TD],
                 diff(index_data[:USB])./index_data[2:end,:USB],
                 diff(index_data[:WFC])./index_data[2:end,:WFC]);
T_norm = hcat(daily_returns[:,1]-mean(daily_returns[:,1]), 
              daily_returns[:,2]-mean(daily_returns[:,2]),
              daily_returns[:,3]-mean(daily_returns[:,3]),
              daily_returns[:,4]-mean(daily_returns[:,4]),
              daily_returns[:,5]-mean(daily_returns[:,5]),
              daily_returns[:,6]-mean(daily_returns[:,6]),
              daily_returns[:,7]-mean(daily_returns[:,7]),
              daily_returns[:,8]-mean(daily_returns[:,8]),
              daily_returns[:,9]-mean(daily_returns[:,9]));
sigma = size(T_norm,1)*transpose(T_norm)*T_norm
show(IOContext(STDOUT), "text/plain", sigma)

9×9 DataArrays.DataArray{Float64,2}:
 11.141    5.09347  10.9481   7.9007   7.81016  10.474    4.10859  6.15085  6.11595
  5.09347  6.10928   5.67271  4.38973  4.29132   5.50456  4.42736  3.48702  3.3516 
 10.9481   5.67271  12.0995   8.45402  8.34689  11.2926   4.72256  6.50711  6.51138
  7.9007   4.38973   8.45402  7.37405  6.12449   8.72175  3.65295  5.00749  4.83638
  7.81016  4.29132   8.34689  6.12449  6.67676   8.09324  3.43315  5.08592  4.82634
 10.474    5.50456  11.2926   8.72175  8.09324  12.3576   4.6782   6.52446  6.38642
  4.10859  4.42736   4.72256  3.65295  3.43315   4.6782   3.98097  2.89125  2.81051
  6.15085  3.48702   6.50711  5.00749  5.08592   6.52446  2.89125  4.6497   4.08472
  6.11595  3.3516    6.51138  4.83638  4.82634   6.38642  2.81051  4.08472  4.66212

The correlation matrix is then calculated as $$\Omega = (D_{\sigma})^{-1}\Sigma(D_{\sigma})^{-1}$$ 
where $D_{\sigma}$ is the diagonal matrix with diagonal values equivalent to the standard deviations of each random variable $X_{i}$   

In [99]:
d_sigma_inverse = inv(Diagonal([sqrt(sigma[1,1]); 
                                sqrt(sigma[2,2]);
                                sqrt(sigma[3,3]);
                                sqrt(sigma[4,4]);
                                sqrt(sigma[5,5]);
                                sqrt(sigma[6,6]);
                                sqrt(sigma[7,7]);
                                sqrt(sigma[8,8]);
                                sqrt(sigma[9,9])]))
omega = d_sigma_inverse*sigma*d_sigma_inverse
show(IOContext(STDOUT), "text/plain", omega)

9×9 DataArrays.DataArray{Float64,2}:
 1.0       0.617387  0.942957  0.871668  0.905556  0.892651  0.616932  0.854597  0.848616
 0.617387  1.0       0.6598    0.654019  0.671912  0.633519  0.897749  0.654255  0.628009
 0.942957  0.6598    1.0       0.895007  0.928662  0.923515  0.680455  0.867544  0.866956
 0.871668  0.654019  0.895007  1.0       0.872839  0.913656  0.674212  0.855174  0.824851
 0.905556  0.671912  0.928662  0.872839  1.0       0.890988  0.665909  0.912797  0.865054
 0.892651  0.633519  0.923515  0.913656  0.890988  1.0       0.666985  0.860725  0.841391
 0.616932  0.897749  0.680455  0.674212  0.665909  0.666985  1.0       0.672014  0.652378
 0.854597  0.654255  0.867544  0.855174  0.912797  0.860725  0.672014  1.0       0.87732 
 0.848616  0.628009  0.866956  0.824851  0.865054  0.841391  0.652378  0.87732   1.0     

For log returns, the process is the same, as shown below the figures do not differ much from those obtained with percentage returns as it is expected for small periods of time.

In [100]:
daily_log_returns = hcat(diff(log(index_data[:BAC])), diff(log(index_data[:BNS])),
    diff(log(index_data[:C])), diff(log(index_data[:GS])),
    diff(log(index_data[:JPM])), diff(log(index_data[:MS])),
    diff(log(index_data[:TD])), diff(log(index_data[:USB])), diff(log(index_data[:WFC])));
T_norm_2 = hcat(daily_log_returns[:,1]-mean(daily_log_returns[:,1]), 
              daily_log_returns[:,2]-mean(daily_log_returns[:,2]),
              daily_log_returns[:,3]-mean(daily_log_returns[:,3]),
              daily_log_returns[:,4]-mean(daily_log_returns[:,4]),
              daily_log_returns[:,5]-mean(daily_log_returns[:,5]),
              daily_log_returns[:,6]-mean(daily_log_returns[:,6]),
              daily_log_returns[:,7]-mean(daily_log_returns[:,7]),
              daily_log_returns[:,8]-mean(daily_log_returns[:,8]),
              daily_log_returns[:,9]-mean(daily_log_returns[:,9]));
sigma_2 = size(T_norm_2,1)*transpose(T_norm_2)*T_norm_2
show(IOContext(STDOUT), "text/plain", sigma_2)

9×9 DataArrays.DataArray{Float64,2}:
 11.0219   5.07725  10.8214   7.80128  7.77423  10.3331   4.06894  6.08914  6.09098
  5.07725  6.16816   5.65022  4.37001  4.30747   5.47519  4.44102  3.48281  3.35734
 10.8214   5.65022  11.94     8.33007  8.29818  11.1098   4.66081  6.43149  6.46893
  7.80128  4.37001   8.33007  7.28173  6.08043   8.5852   3.61577  4.95274  4.8002 
  7.77423  4.30747   8.29818  6.08043  6.69005   8.02143  3.4257   5.0613   4.82962
 10.3331   5.47519  11.1098   8.5852   8.02143  12.1466   4.62175  6.44351  6.33143
  4.06894  4.44102   4.66081  3.61577  3.4257    4.62175  3.97594  2.87414  2.79597
  6.08914  3.48281   6.43149  4.95274  5.0613    6.44351  2.87414  4.61325  4.06988
  6.09098  3.35734   6.46893  4.8002   4.82962   6.33143  2.79597  4.06988  4.65477

In [102]:
d_sigma_inverse_2 = inv(Diagonal([sqrt(sigma_2[1,1]); 
                                sqrt(sigma_2[2,2]);
                                sqrt(sigma_2[3,3]);
                                sqrt(sigma_2[4,4]);
                                sqrt(sigma_2[5,5]);
                                sqrt(sigma_2[6,6]);
                                sqrt(sigma_2[7,7]);
                                sqrt(sigma_2[8,8]);
                                sqrt(sigma_2[9,9])]))
omega_2 = d_sigma_inverse_2*sigma_2*d_sigma_inverse_2
show(IOContext(STDOUT), "text/plain", omega_2)

9×9 DataArrays.DataArray{Float64,2}:
 1.0       0.615775  0.943308  0.870803  0.905344  0.893048  0.614656  0.853932  0.850373
 0.615775  1.0       0.658394  0.65206   0.670548  0.632548  0.896778  0.652902  0.626568
 0.943308  0.658394  1.0       0.893366  0.928466  0.922524  0.676455  0.866574  0.867723
 0.870803  0.65206   0.893366  1.0       0.871169  0.912862  0.671991  0.854523  0.824505
 0.905344  0.670548  0.928466  0.871169  1.0       0.889834  0.664224  0.911052  0.865464
 0.893048  0.632548  0.922524  0.912862  0.889834  1.0       0.665056  0.860777  0.842024
 0.614656  0.896778  0.676455  0.671991  0.664224  0.665056  1.0       0.671095  0.649924
 0.853932  0.652902  0.866574  0.854523  0.911052  0.860777  0.671095  1.0       0.878271
 0.850373  0.626568  0.867723  0.824505  0.865464  0.842024  0.649924  0.878271  1.0     

**(ii) Compute the sample covariance matrix of the weekly percentage returns of the indices, and the corresponding sample correlation matrix. Compute the sample covariance and correlation matrices for weekly log returns, and compare them with the corresponding matrices for weekly percentage returns.**

In [105]:
using TimeSeriesIO
using TimeSeriesResampler
ta = TimeArray(index_data, colnames=[:BAC, :BNS, :C, :GS, :JPM, :MS, :TD, :USB, :WFC], timestamp=:Date)
weekly = ohlc(resample(ta, Dates.Week(1)))

30x36 TimeSeries.TimeArray{Float64,2,Date,Array{Float64,2}} 2016-01-04 to 2016-07-25
│            │ BAC.Open │ BAC.High │ BAC.Low │ BAC.Close │ BNS.Open │ BNS.High │
├────────────┼──────────┼──────────┼─────────┼───────────┼──────────┼──────────┤
│ 2016-01-04 │ 16.1768  │ 38.1836  │ 50.5166 │ 174.4288  │ 61.87    │ 30.5877  │
│ 2016-01-11 │ 15.0741  │ 36.4689  │ 46.2879 │ 163.1934  │ 57.2118  │ 27.6533  │
│ 2016-01-18 │ 14.0205  │ 35.0128  │ 41.4368 │ 154.4198  │ 55.4419  │ 25.5157  │
│ 2016-01-25 │ 12.7603  │ 35.5205  │ 39.0755 │ 148.807   │ 54.129   │ 24.272   │
│ 2016-02-01 │ 13.7449  │ 38.7488  │ 42.0222 │ 157.2064  │ 57.241   │ 25.1292  │
│ 2016-02-08 │ 12.0809  │ 38.2986  │ 37.4025 │ 146.9656  │ 54.9848  │ 22.1578  │
│ 2016-02-15 │ 12.0612  │ 37.992   │ 38.4708 │ 146.7392  │ 56.745   │ 23.1841  │
│ 2016-02-22 │ 12.3467  │ 38.8159  │ 39.0941 │ 146.5127  │ 56.9589  │ 23.9856  │
│ 2016-02-29 │ 12.3271  │ 38.3177  │ 38.4313 │ 147.8811  │ 54.7514  │ 24.142   │
│ 2016-03-07 │ 13.3722  

**(iii) Compute the sample covariance matrix of the monthly percentage returns of the indices, and the corresponding sample correlation matrix. Compute the sample covariance and correlation matrices for monthly log returns, and compare them with the corresponding matrices for monthly percentage returns.**

**(iv) Comment on the differences between the sample covariance and correlation matrices for daily, weekly and monthly returns.** 

--------------------
**(12) Consider a market made of cash and an asset with spot price 40. The continuously compounded risk-free rate is 2%. In four months, the price of the stock will be either 45 or 35.**

**(i) What is the payoff matrix M in three months?**

The future value of cash is $e^{0.02\times 0.25} = 1.00501252$ and

$$
M=\begin{pmatrix} 
      1.00501252 & 1.00501252 \\ 
      45 & 35
   \end{pmatrix}\\
$$

**(ii) Is the market complete, i.e., is the matrix M nonsingular?**

$$
\begin{align}
det(M) =&  1.00501252\times 35 - 1.00501252\times 45 \\
=& 35.1754382 - 45.2255634\\ 
=& -10.0501252 \neq 0\\ 
\implies& \text{the matrix M is nonsingular and the market is complete.}
\end{align}
$$

**(iii) How do you replicate a four months at-the-money call option on this asset, using the cash and the underlying asset?**

The payoff of a 4 months at-the-money call option if the spot price is 40 is $$P\left(\frac{4}{12}\right) = max\left(S\left(\frac{4}{12}\right) - 40, 0\right)$$

For each given state, the price $S\left(\frac{4}{12}\right)$ is:
$$
\omega^1 : S\left(\frac{4}{12}\right) = 45, P\left(\frac{4}{12}\right) = 5\\
\omega^2 : S\left(\frac{4}{12}\right) = 35, P\left(\frac{4}{12}\right) = 0\\
$$

Therefore, the value of the portfolio that replicates the call option is:
$$
\begin{align}
V\left(\frac{4}{12}\right) =& (\Theta_{1} && \Theta_{2})\begin{pmatrix} 
      1.00501252 & 1.00501252 \\ 
      45 & 35
   \end{pmatrix} = (5 && 0)\\
\end{align}
$$

$$
\implies
\begin{pmatrix} 
      \Theta_{1} \\
      \Theta_{2}
\end{pmatrix} = 
\begin{pmatrix} 
      -17.4127 \\
      0.5
\end{pmatrix}
$$


The call option can be replicated by shorting -17.4127 in cash and buying 0.5 units of the asset.

----------------
**(13) In six months, the price of an asset with spot price 42 will be either 35, 38, 40, 43, 45 or 50. Consider a market made of cash, the stock, and a six months at-the-money call and a six months at-the-money put option on one share of the stock. The risk-free interest rate is constant equal to 2%.**

**(i) What is the payoff matrix in six months?**

The future value of cash is $e^{0.02\times 0.5} = 1.0100502$, the possible values of the call option are $C\left(\frac{1}{2}\right) = (0 \ 0 \ 0 \ 1 \ 3 \ 8)$, the possible values of the put option are $P\left(\frac{1}{2}\right) = (7 \ 4 \ 2 \ 0 \ 0 \ 0)$

$$
M=\begin{pmatrix} 
      1.0100502 & 1.0100502 & 1.0100502 & 1.0100502 & 1.0100502 & 1.0100502\\ 
      35 & 38 & 40 & 43 & 45 & 50\\ 
      0 & 0 & 0 & 1 & 3 & 8\\ 
      7 & 4 & 2 & 0 & 0 & 0\\ 
   \end{pmatrix}\\
$$

**(ii) Is this market complete?**

No, the rank of the payoff matrix is 3 which is less than the number of possible states [Theorem 3.3]

**(iii) Are all the securities in this market non-redundant?**

No, the rank of the payoff matrix is less than the number of possible states.
For example, the first row can be expressed as a linear combination of the second, third and fourth rows:
$S_{1} = \frac{e^{0.01}}{42}(S_{2} - S_{3} + S_{4})$